File: MATLAB_VARIANT/fbi.m
function [th, la, ff, hh] = fbi(f, h, f_, n, m, p, n_, d)
%FBI  Term-by-term solution of the Francis-Byrnes-Isidori (FBI) PDE for nonlinear regulation
%Author: gpertin, KAIST
%Copyright (rewrite) 2025.  Original algorithm: A. J. Krener

    df = [1, d];                              
    %degree-1 solution
    Fx  = f(:, 1:n);                          %dF/dx        (N x N)
    Fu  = f(:, n+1:n+m);                      %dF/du        (N x M)
    Fxb = f(:, n+m+1:n+m+n_);                 %dF/dx_       (N x N_)
    Hx  = h(:, 1:n);                          %dH/dx        (P x N)
    Hu  = h(:, n+1:n+m);                      %dH/du        (P x M)
    Hxb = h(:, n+m+1:n+m+n_);                 %dH/dx_       (P x N_)
    A_  = f_(:, 1:n_);                        %linear exosystem  x_' = A_ x_

    %Combined linear plant/output map and the selector onto the x-block
    M_  = [Fx, Fu; Hx, Hu];                   %(N+P) x (N+M)
    Sel = [eye(n), zeros(n,m); zeros(p, n+m)];%(N+P) x (N+M), picks the x part

    %Degree 1. Solve, for T (N x N_) and L (M x N_):
    %     Fx*T + Fu*L - T*A_ = -Fxb        (the F rows)
    %     Hx*T + Hu*L        = -Hxb        (the H rows)
    %Using (P*Z*Q) row-stacked = Z row-stacked * kron(P', Q), this is one
    %linear system in the stacked unknown [T;L]
    Op  = kron(M_.',  eye(n_)) - kron(Sel.', A_);
    rhs = -reshape([Fxb; Hxb].', 1, (n+p)*n_);
    sol = rhs * pinv(Op);
    th  = reshape(sol(1, 1:n*n_),            n_, n).';   %N  x N_
    la  = reshape(sol(1, n*n_+1:(n+m)*n_),   n_, m).';   %M  x N_

    %Degrees 2-D. Same linear operator (now acting on degree-k blocks),
    %with a right-hand side assembled from the lower-degree solution
    for k = 2:d
        n_k = fbipoly.crd(n_, k);             %degree-k monomials in x_
        %Linear operator at degree k
        LF = fbipoly.dd(eye(n_k), [n_k, n_], k, A_, [n_, n_], 1, k);
        Op = kron(M_.', eye(n_k)) - kron(Sel.', LF);
        %RHS term 1
        data_hi = fbipoly.prt([f; h], [n,n; p,m; 0,n_], df, [2, k]);
        g_sub   = [th; la; eye(n_), zeros(n_, fbipoly.crdsum(n_, [2, k-1]))];
        comp    = fbipoly.compose(data_hi, [n,n; p,m; 0,n_], [2, k], ...
                                  g_sub,   [n,0; m,0; n_,n_], [1, k-1], k);
        %RHS term 2
        f_hi  = fbipoly.prt(f_, [n_, n_], df, [1, k]);
        deriv = [fbipoly.dd(th, [n, n_], [1, k-1], f_hi, [n_, n_], [1, k], k); ...
                 zeros(p, n_k)];
        rhs = deriv - comp;                           % (N+P) x n_k
        rhs = reshape(rhs.', 1, n_k*(n+p));           % row vector
        %Solve  rhs = sol * Op  (exact if square & nonsingular, else lsq).
        if size(Op,1) == size(Op,2)
            sol = rhs / Op;
        else
            sol = rhs * pinv(Op);
        end
        sol = reshape(sol, n_k, n+m).';               
        th  = [th, sol(1:n,     :)];                  
        la  = [la, sol(n+1:n+m, :)];                  
    end

    %Transverse coordinates  z = x - TH(x_),  v = u - LA(x_),  x_ = x_ .
    ph = [eye(n+m+n_), zeros(n+m+n_, fbipoly.crdsum(n+m+n_, [2, d]))];
    thall = fbipoly.compose(th, [n 0 0; 0 0 n_].', df, ...
                            [zeros(n_, n+m), eye(n_)], [0 0 n_; n m n_].', 1, df);
    ph(1:n, :) = ph(1:n, :) + thall;
    laall = fbipoly.compose(la, [0 m 0; 0 0 n_].', df, ...
                            [zeros(n_, n+m), eye(n_)], [0 0 n_; n m n_].', 1, df);
    ph(n+1:n+m, :) = ph(n+1:n+m, :) + laall;
    nf  = [n 0 0; n m n_].';
    nh  = [p 0 0; n m n_].';
    nth = [n n_];
    nf_ = [n_ n_];
    nph = [n m n_; n m n_].';
    ff = fbipoly.compose(f, nf, df, ph, nph, df, df);
    hh = fbipoly.compose(h, nh, df, ph, nph, df, df);
    temp = fbipoly.dd(th, nth, df, f_, nf_, df, df);
    ff = ff - fbipoly.compose(temp, [n 0 0; 0 0 n_].', df, ...
                              [zeros(n_, n+m), eye(n_)], [0 0 n_; n m n_].', 1, df);
end

File: src/NSTJAX/NSTJAX_suite/fbi.py
"""
Author: gpertin, KAIST
Solves the Francis-Byrnes-Isidori equations in JAX
Based on Krener's Nonlinear Systems Toolbox
"""

from functools import partial
import jax
import jax.numpy as jnp
from NSTJAX.NSTJAX_suite.polylib import crd, unpack, zero_field, compose, ddmul, lie_operator, comp_operator

def _vec(X):
    #Column major vectorization
    return X.T.reshape(-1)

def _unvec(v, r, c):
    #Inverse of _vec for an r by c matrix
    return v.reshape(c, r).T

def _solve(Op, rhs):
    #Square -> direct solve, otherwise least squares.
    if Op.shape[0] == Op.shape[1]:
        W = jnp.linalg.solve(Op, rhs)
    else:
        W = jnp.linalg.lstsq(Op, rhs, rcond=None)[0]
    return W

def _operator(M_, Sel, LF):
    #Degree k regulator operator for the equation M_ W - Sel W LF = RHS
    nk = LF.shape[0]
    eye = jnp.eye(nk, dtype=M_.dtype)
    return jnp.kron(eye, M_) - jnp.kron(LF.T, Sel)

#Partial compiled to improve speed
@partial(jax.jit, static_argnums=(3, 4, 5, 6, 7, 8))
def fbi(f, h, fexo, n, m, p, n_, d, disc=False):
    nsum = n + m + n_
    dt = f.dtype
    #Linear parts
    F1 = f[:, :nsum]
    H1 = h[:, :nsum]
    Fxb = F1[:, n + m:nsum]
    Hxb = H1[:, n + m:nsum]
    A_ = fexo[:, :n_]
    #Combined linear plant output map and the selector onto the x block
    M_ = jnp.block([[F1[:, :n], F1[:, n:n + m]],
                    [H1[:, :n], H1[:, n:n + m]]])
    Sel = jnp.block([[jnp.eye(n, dtype=dt), jnp.zeros((n, m), dt)],
                     [jnp.zeros((p, n + m), dt)]])
    #Degree 1, the linear regulator equations
    Op = _operator(M_, Sel, A_)
    rhs = -_vec(jnp.concatenate([Fxb, Hxb], axis=0))
    W = _unvec(_solve(Op, rhs), n + m, n_)
    th = W[:n, :]
    la = W[n:, :]
    #Degrees 2..d, same operator with the lower degree right hand side
    fh = jnp.concatenate([f, h], axis=0)
    fh_field = unpack(fh, nsum, 1, d)
    fexo_field = unpack(fexo, n_, 1, d)
    for k in range(2, d + 1):
        n_k = crd(n_, k)
        LF = comp_operator(A_, n_, k) if disc else lie_operator(A_, n_, k)
        Op = _operator(M_, Sel, LF)
        #Term 1, composition cross term (discrete) or derivative cross term (continuous)
        th_field = unpack(th, n_, 1, k - 1)
        la_field = unpack(la, n_, 1, k - 1)
        deriv = (compose(th_field, n_, fexo_field, n_, k)[k] if disc
                 else ddmul(th_field, fexo_field, n_, k)[k])
        #Term 2, the nonlinear part of [f; h] composed with (th, la, x_)
        data_hi = zero_field(n + p, nsum, k, dtype=dt)
        for kk in range(2, k + 1):
            data_hi[kk] = fh_field[kk]
        g_sub = zero_field(nsum, n_, k, dtype=dt)
        for deg in range(1, k):
            g_sub[deg] = g_sub[deg].at[:n].set(th_field[deg])
            g_sub[deg] = g_sub[deg].at[n:n + m].set(la_field[deg])
        g_sub[1] = g_sub[1].at[n + m:nsum].set(jnp.eye(n_, dtype=dt))
        comp = compose(data_hi, nsum, g_sub, n_, k)[k]
        #Assemble and solve
        rhs_F = deriv - comp[:n]
        rhs_H = -comp[n:]
        rhs = _vec(jnp.concatenate([rhs_F, rhs_H], axis=0))
        W = _unvec(_solve(Op, rhs), n + m, n_k)
        th = jnp.concatenate([th, W[:n, :]], axis=1)
        la = jnp.concatenate([la, W[n:, :]], axis=1)

    return th, la
